test/e2e/app/app.go

```diff
@@ -308,12 +315,39 @@ func (app *Application) PrepareProposal(
 		if !vote.SignedLastBlock || len(vote.VoteExtension) == 0 {
 			continue
 		}
+		// We unconditionally provide vote extension signatures. It's up to the app to use them or not
+		if len(vote.ExtensionSignature) == 0 {
+			panic("ABCI passed the application a vote extension with empty signature")
+		}
+
+		// Reconstruct vote extension's signed bytes...
+		cve := cmtproto.CanonicalVoteExtension{
+			Extension: vote.VoteExtension,
+			Height:    req.Height,
+			Round:     int64(req.LocalLastCommit.Round),
+			ChainId:   app.chanID,
+		}
+		extSignBytes, err := protoio.MarshalDelimited(&cve)
+		if err != nil {
+			panic(err)
+		}
+
+		//... and verify
+		valAddr := crypto.Address(vote.Validator.Address)
+		pub, ok := app.currentValidators[valAddr.String()]
+		if !ok {
+			panic("PrepareProposal: received vote from unknown validator")
+		}
+
+		if !pub.VerifySignature(extSignBytes, vote.ExtensionSignature) {
+			panic("PrepareProposal: received vote with invalid signature")
+		}
```

Comment:
What would the recommended behaviour be for correct applications here? I assume it's not to panic, right? (Just asking preemptively for app developers' sake - I can see how this is useful for picking up problems in our E2E tests)

Reply:
So, my understanding is that the signature we're verifying is local (it comes from local CometBFT, which is supposed to have verified it itself when it received the info). A real app could even skip this verification in `PrepareProposal` under the argument it trusts CometBFT has verified it.

The missing piece of this PR (hence it's still a draft) is shipping all necessary info in special transactions so that `ProcessProposal` can also verify (which is where a real app would verify). In that case it wouldn't be a panic, but a `ResponseProcessProposal` with `reject`

Reply:
In the case of the app, which verifies vote exts "injected" by the proposer, if any such verification fails, rather if more than 2/3 voting power fails, then it would REJECT the proposal.

Reply:
I didn't get this last comment. The proposal contains extensions from 2/3 of the voting power from the previous round, out of which 1/3+1 is guaranteed to be from correct validators. What do you mean by "if more than 2/3 voting power fails"?
Maybe I am not understanding how you plan on using it. Is there anything you could share?

Reply:
> The proposal contains extensions from 2/3 of the voting power from the previous round, out of which 1/3+1 is guaranteed to be from correct validators.

Yes, given to the proposer during `PrepareProposal`. The remaining validators that receive the manually injected vote extensions during `ProcessProposal` have to ensure they get at least 2/3 from the proposer.